### PR TITLE
DOCSP-13829  authentication mechanism

### PR DESCRIPTION
**Pull request by kyuan-mongodb**
## Pull Request Info

Added Fundamentals > Authentication Mechanisms page

Utilized snippets from Go driver repo: https://github.com/mongodb/mongo-go-driver/blob/master/mongo/client_examples_test.go 

Based the format off of the [Node](https://docs.mongodb.com/drivers/node/fundamentals/authentication/mechanisms/) and [Java](https://docs.mongodb.com/drivers/java/sync/current/fundamentals/auth/) Auth Mechanism pages.

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-13829

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/golang/docsworker-xlarge/DOCSP-13829-AuthMechanisms/fundamentals/auth/

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

### CODE REVIEW

File: source/fundamentals/auth.txt [hunk 1]
```diff
@@ -18,14 +21,251 @@ confirm identity and establish trust to ensure security.
...
 
-To authenticate using ``Kerberos`` or ``LDAP``, see the
-:doc:`Enterprise Authentication Mechanisms guide </fundamentals/enterprise-auth>`.
 For more information on establishing a connection to your MongoDB cluster,
 read our :doc:`Connection Guide </fundamentals/connection>`.
+
+Mechanisms
+----------
+
+Authenticate a mechanism by specifying the authentication options
```
Comment (terakilobyte) [inline source/fundamentals/auth.txt:37]: nit: What is `authenticate a mechanism`?

File: source/fundamentals/auth.txt [hunk 2]
```diff
@@ -18,14 +21,251 @@ confirm identity and establish trust to ensure security.
...
+through an ``Credential`` type. The ``Credential`` gets passed  to
```
Comment (terakilobyte) [inline source/fundamentals/auth.txt:38]:
```suggestion
through a ``Credential`` type. The ``Credential`` gets passed  to
```

**Review by terakilobyte — CHANGES_REQUESTED** [review]
It looks like this is nearly identical to the Java page (minus the go specifics), but removes one of the headings. I think it worked well.

File: source/fundamentals/auth.txt [hunk 3]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+     - Versions
+
+   * - ``SCRAM-SHA-256``
+     - MongoDB 4.0 and later   
+   
+   * - ``SCRAM-SHA-1``
+     - MongoDB 3.0, 3.2, 3.4, and 3.6
+
+   * - ``MONGODB-CR``
+     - MongoDB 3.6 and earlier
```
Comment (terakilobyte) [inline source/fundamentals/auth.txt:75]: I believe this should be 2.6 and earlier.

**Review by terakilobyte — APPROVED** [review]
LGTM with one correction needed.

File: source/fundamentals/auth.txt (hunk 3 above)
Comment (kyuan-mongodb) [reply source/fundamentals/auth.txt:75]: hm, I found the mention in the [server manual release notes ](https://docs.mongodb.com/manual/release-notes/3.6/#security) and it says it was depreciated as of 3.6 -- maybe the edit is "earlier than MongoDB 3.6"?

**Comment by kyuan-mongodb** [comment]
Hi @kevinAlbs, This PR is ready for review.

File: source/fundamentals/auth.txt (hunk 3 above)
Comment (benjirewis) [reply source/fundamentals/auth.txt:75]: Based on [`chooseAuthMechanism`](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/auth/default.go#L71-L73) it looks like we pick `SCRAM-SHA-256` if it's supported, falling back to `SCRAM-SHA-1` if not, and finally to `MONGODB-CR` if neither is supported.

File: source/fundamentals/auth.txt [hunk 4]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+
+   ``SCRAM-SHA-256`` is the default authentication method for MongoDB starting
+   in MongoDB 4.0.
+
+``SCRAM-SHA-256`` is a salted challenge-response authentication mechanism
+(SCRAM) that uses your username and password, encrypted with the ``SHA-256``
+algorithm, to authenticate your user.
+
+To specify the ``SCRAM-SHA-256`` authentication mechanism, assign the
+``AuthMechanism`` option the value ``SCRAM-SHA-256``:
```
Comment (benjirewis) [inline source/fundamentals/auth.txt:109]:
```suggestion
``AuthMechanism`` option the value ``"SCRAM-SHA-256"``:  
```

File: source/fundamentals/auth.txt [hunk 5]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+~~~~~~~~~~~~~~
+
+``MONGODB-CR`` is a challenge-response authentication mechanism that uses your
+username and password to authenticate your user. This authentication
+mechanism was deprecated starting in MongoDB 3.6 and is no longer
+supported as of MongoDB 4.0.
+
+You cannot specify this method explicitly; refer to the fallback provided
+by the :ref:`default authentication mechanism <default-auth-mechanism>` to
+connect using ``MONGODB-CR``.
```
Comment (benjirewis) [inline source/fundamentals/auth.txt:166]: I believe you can specify `MONGODB-CR` in the Go driver by passing `"MONGODB-CR"` as the `AuthMechanism`.

File: source/fundamentals/auth.txt [hunk 6]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+option the value of your ``sessionToken``:
+
+.. code-block:: go
+   :emphasize-lines: 7-9
+
+   var sessionToken string
+   assumeRoleCredential := options.Credential{
+      AuthMechanism: "MONGODB-AWS",
+      AuthSource: "<authenticationDb>",
+      Username:      accessgKeyID,
```
Comment (benjirewis) [inline source/fundamentals/auth.txt:218]:
```suggestion
      Username:      accessKeyID,
```

File: source/fundamentals/auth.txt [hunk 7]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+
+- ``tlsCAFile`` which contains either a single or a bundle of certificate authorities to trust when making a TLS connection
+- ``tlsCertificateKeyFile`` which references the path to the client certificate file or the client private key file
+
+To specify the ``X.509`` authentication mechanism, perform the
+following:
+
+- Assign the ``tlsCAFile`` the path to its file in the connection string
+- Assign the ``tlsCertificateKeyFile`` the path to its file in the connection string
+- Assign the ``AuthMechanism`` option the value ``X.509``
```
Comment (benjirewis) [inline source/fundamentals/auth.txt:245]:
```suggestion
- Assign the ``AuthMechanism`` option the value ``"MONGODB-X509"``
```

**Review by benjirewis — COMMENTED** [review]
Looking good; have some suggestions.

The Go driver also offers `GSSAPI` and `PLAIN` authentication mechanisms that might be worth mentioning.

File: source/fundamentals/auth.txt [hunk 8]
```diff
@@ -10,22 +10,254 @@ Authentication Mechanisms
...
+ensure security in the driver and server before connecting.
+
+If you want to establish a connection to your MongoDB cluster without an
+authentication mechanism, see our :doc:`Connection Guide </fundamentals/connection>`.
+
+Supported Mechanisms
+--------------------
+
+The Go Driver establishes a connection with an authentication mechanism
+through a `Client <https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.7.1/mongo#Client>`__
```
[the remark this answers is not in the document]
Comment (kyuan-mongodb) [reply source/fundamentals/auth.txt:28]: We usually use the version that corresponds the one we're publishing -- this accounts for changes made in future releases that may contradict what we currently have written.

File: source/fundamentals/auth.txt (hunk 3 above)
Comment (kyuan-mongodb) [reply source/fundamentals/auth.txt:75]: The link you provided contains `/x`, which I was told is experimental and not for public use.

File: source/fundamentals/auth.txt [hunk 9]
```diff
@@ -72,7 +73,7 @@ mechanisms depending on what MongoDB versions your server supports:
      - MongoDB 3.0, 3.2, 3.4, and 3.6
 
    * - ``MONGODB-CR``
-     - MongoDB 3.6 and earlier
+     - MongoDB 3.6 and earlier if the server doesn't support``SCRAM-SHA-1``
```
Comment (terakilobyte) [inline source/fundamentals/auth.txt:76]: This still isn't correct. This table is for the *default* it looks like. [MongoDB 3.0](https://docs.mongodb.com/v3.0/core/security-mongodb-cr/) changed the default from `MongoDB-CR` to `SCRAM-*`.

**Review by terakilobyte — CHANGES_REQUESTED** [review]

File: source/fundamentals/auth.txt (hunk 3 above)
Comment (benjirewis) [reply source/fundamentals/auth.txt]: As discussed offline: most of our authentication logic is "unstable" API and is therefore under `/x`. I wouldn't point to code under `/x` in documentation, but I was just linking to that code here to explain how the Go driver actually picks the "default" auth mechanism.

**Comment by benjirewis** [comment]
As discussed offline: while I do believe you _can_  explicitly specify `MONGODB-CR` as an auth mechanism, our documentation about its deprecation and eventual removal was slightly incorrect. A [fix](https://github.com/mongodb/mongo-go-driver/pull/718) is in review now.

**Review by benjirewis — APPROVED** [review]
LGTM! Nice work.

**Review by terakilobyte — APPROVED** [review]
lgtm

**Pull request merged**